db/migrate/20250313020004_add_not_null_constraints_to_participants.rb

```diff
@@ -0,0 +1,15 @@
+class AddNotNullConstraintsToParticipants < ActiveRecord::Migration[7.1]
+  def change
+    # update any existing NULL values, just in case
+    Participant.where(email: nil).update_all(email: 'INVALID')
+    Participant.where(crypted_password: nil).update_all(crypted_password: 'INVALID')
```

Comment:
Strictly speaking, migrations shouldn’t use models, because this migration has to work with _every_ future version of the model code (but with the schema frozen at this point in the migration chain). In practice, however, I really doubt this will cause any actual problems!

Reply:
That's fair. I agree with you - I think the likelihood that the `Participant` model goes away are _very_ low.